### PR TITLE
GraphClient for the high level REST client and associated tests

### PR DESCRIPTION
**Pull request by markharwood**
Documentation yet to be done

### CODE REVIEW

**Comment by markharwood** [comment]
test this please

File: x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/graph/Connection.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+import static org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+
+/**
+ * A Connection links exactly two {@link Vertex} objects. The basis of a 
+ * connection is one or more documents have been found that contain
+ * this pair of terms and the strength of the connection is recorded
+ * as a weight.
+ */
+public class Connection {
+    Vertex from;
```
Comment (nik9000) [inline x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/graph/Connection.java]: I think it'd be nice if these were private. We've been trying to be more "standard-java" in the rest client.

File: x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/graph/Connection.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+            this.fromIndex = fromIndex;
+            this.toIndex = toIndex;
+            this.weight = weight;
+            this.docCount = docCount;
+        }
+        public Connection resolve(List<Vertex> vertices) {            
+            return new Connection(vertices.get(fromIndex), vertices.get(toIndex), weight, docCount);
+        }
+        
+        static UnresolvedConnection fromXContent(XContentParser parser) throws IOException {
```
Comment (nik9000) [inline x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/graph/Connection.java]: I think `ObjectParser` would be easier to review but this looks right so I'm fine with it.

File: x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/graph/GraphExploreRequest.java [hunk 3]
```diff
@@ -330,6 +359,41 @@ void writeTo(StreamOutput out) throws IOException {
         }
 
     }
-    
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        
+        builder.startObject("controls"); //=== Start Controls
```
Comment (nik9000) [inline x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/graph/GraphExploreRequest.java]: I've borrowed a style from dakrone that looks like
```
builder.startObject("controls");
{
  if (sampleSize != ... ) {
    builder.field("sample_size", sampleSize);
  }
  ...
}
builder.endObject();
```

It sort of replaces the need for `//=== Start Controls`.

File: x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/graph/GraphExploreRequest.java [hunk 4]
```diff
@@ -330,6 +359,41 @@ void writeTo(StreamOutput out) throws IOException {
...
+            builder.endObject();
+        }
+        builder.field("use_significance", useSignificance);
+        if (returnDetailedInfo) {
+            builder.field("return_detailed_stats", returnDetailedInfo);
+        }
+        builder.endObject(); //=== End controls
+
+        for (Hop hop : hops) {
+            if (hop.parentHop != null) {
```
Comment (nik9000) [inline x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/graph/GraphExploreRequest.java]: They look like a chain!  Woah!

File: x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/graph/GraphExploreResponse.java [hunk 5]
```diff
@@ -198,5 +216,59 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
+        GraphExploreResponse result = new GraphExploreResponse();  
+        result.vertices = new HashMap<>();
+        result.connections = new HashMap<>();
+        ensureExpectedToken(Token.START_OBJECT, parser.nextToken(), parser::getTokenLocation);
+        parser.nextToken();
+        String currentFieldName = parser.currentName();
+        List<Vertex> vertices = new ArrayList<>();
+        List<ShardSearchFailure> failures = new ArrayList<>();
+        List<Connection.UnresolvedConnection> unresolvedConnections = new ArrayList<>();
+        for (Token token = parser.nextToken(); token != Token.END_OBJECT; token = parser.nextToken()) {
```
Comment (nik9000) [inline x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/graph/GraphExploreResponse.java]: This one also looks right when I scan it but'd be easier to review as `ObjectParser`.

**Comment by hub-cap** [comment]
I have just committed #32596 which changes the location of the xpack clients. Its very likely your tests wont compile now that the xpack portion of `client.xpack().yourCommercialClient()` is removed.

File: x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/graph/GraphExploreResponse.java (hunk 5 above)
Comment (markharwood) [reply x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/graph/GraphExploreResponse.java]: I may be missing something but isn't ObjectParser more geared towards parsing "request" objects? Requests tend to have existing "Builder" classes with setter methods used by clients to express criteria and ObjectParser relies on registering setter methods.

Response objects tend to be immutable data objects and adding "setter" methods to accommodate the use of ObjectParser is therefore perhaps the wrong approach?

Comment (nik9000) [reply x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/graph/GraphExploreResponse.java]: `ConstructingObjectParser` can mostly get the job done without needing any extra builders. But sometimes builders are easier to understand. The idea with using these parsers is that they are a ton easier to review for correctness, especially that they properly ignore new fields. It isn't required, but it does make it easier to be sure that it is all rigged up right.

**Comment by markharwood** [comment]
Many thanks for the review, @nik9000 . I've switched the code over to ObjectParser.

Good to go @nik9000 @hub-cap ?

File: docs/java-rest/high-level/supported-apis.asciidoc [hunk 6]
```diff
@@ -217,3 +217,15 @@ The Java High Level REST Client supports the following Watcher APIs:
...
+
+The Java High Level REST Client supports the following Graph APIs:
+
+* <<java-rest-high-x-pack-graph-explore>>
+
+include::graph/explore.asciidoc[]
+
+
+
+
```
Comment (hub-cap) [inline docs/java-rest/high-level/supported-apis.asciidoc]: thats a lot of blank lines!

**Comment by hub-cap** [comment]
if @nik9000 is good then Im good too. I just did a super quick eyeball and didnt see anything insane. :shipit: if its got an approval from him... i was out all last week so im still coming back up to speed :)

**Comment by markharwood** [comment]
Any other budding reviewers you can suggest, @hub-cap ?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java [hunk 7]
```diff
@@ -2598,6 +2601,35 @@ public void testXPackPutWatch() throws Exception {
         request.getEntity().writeTo(bos);
         assertThat(bos.toString("UTF-8"), is(body));
     }
+    
+    public void testXPackGraphExplore() throws Exception {
```
Comment (nik9000) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: I think maybe this should be `testGraphExplore` now that we're removing the `xpack` in lots of places.

File: x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/graph/Connection.java [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    private long docCount;
+
+    public Connection(Vertex from, Vertex to, double weight, long docCount) {
+        this.from = from;
+        this.to = to;
+        this.weight = weight;
+        this.docCount = docCount;
+    }
+
+    void readFrom(StreamInput in, Map<VertexId, Vertex> vertices) throws IOException {
```
Comment (nik9000) [inline x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/graph/Connection.java]: Why not a ctor that takes StreamInput and Map?

**Review by nik9000 — APPROVED** [review]
Left two little picky things but LGTM anyway. Thanks for working through all of my requests!

**Comment by markharwood** [comment]
Many thanks, @nik9000 
I addressed your review comments and will push when I get a green result here.

**Pull request merged**